feat: multiple locations on courses

ap0nia opened the pull request:
## Summary
A continuation of #634

ap0nia committed:
feat: multiple locations on courses

ap0nia commented on apps/antalmanac/src/components/Map/Map.tsx:
```diff
@@ -97,33 +99,33 @@ export default function CourseMap() {
-    }, [setCalendarEvents]);
-
     useEffect(() => {
+        const updateMarkers = () => {
+            setMarkers(getCoursesPerBuilding());
+        };
```
Minor housekeeping to clean this up.

EricPedley replied:
Thx for these comments to explain why these changes are here. Although again, would be better in their own PR that is just for refactoring. A little git hack that might make it easier:

1. selectively stage only the refactor changes to your index
2. commit them as their own atomic commit
3. stash
4. revert the refactor commit
5. checkout to main, make a new branch, then reflog to find the id and cherry-pick that commit back onto your new branch
6. open new PR
7. checkout back to your feature branch and pop the stash back on

Now that I lay it out like that it sounds slightly like a pain in the ass, but if are experienced enough to not need to look up any of the commands it's pretty fast, and probably saves more time during review than it takes during writing.

ap0nia commented on apps/antalmanac/src/components/Map/Map.tsx:
```diff
@@ -97,33 +99,33 @@ export default function CourseMap() {
+    }, []);
 
     useEffect(() => {
+        const updateCalendarEvents = () => {
+            setCalendarEvents(AppStore.getCourseEventsInCalendar());
+        };
```
Housekeeping.

ap0nia committed:
fix: test cases

EricPedley commented on apps/antalmanac/src/components/Map/Map.tsx:
```diff
@@ -41,7 +41,9 @@ interface MarkerContent {
-    const allBuildingCodes = courseEvents.map((event) => event.bldg.split(' ').slice(0, -1).join(' '));
+    const allBuildingCodes = courseEvents
+        .map((event) => event.bldg.map((location) => location.building))
+        .reduce((prev, cur) => prev.concat(cur), []);
```
Can we just replace this reduce with a `flat()` call? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat

ap0nia replied:
Oh, I didn't realize what it was doing; I was just copying the code lol. But I can do you one better with `flatMap`

```suggestion
        courseEvents.flatMap(event => event.bldg.map(location => location.building))
```

EricPedley commented on apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx:
```diff
@@ -88,8 +88,25 @@ interface CommonCalendarEvent extends Event {
+
 export interface CourseEvent extends CommonCalendarEvent {
-    bldg: string; // E.g., ICS 174, which is actually building + room
+    bldg: Location[];
```
Can we rename this to `locations`?

alanchangxyz replied:
looking into this and some of the queries from api also return `bldg` instead of `locations`; going to see if i can also make those changes in this pr or if there's going to be a need for a related backend pr/intermediate formatting

alanchangxyz replied:
update: decided to use intermediate formatting for replacing parts where api gives `bldg` instead of `locations`, yall can do a separate migration for backend later if this needs to become standard across all parts of the stack

EricPedley commented on apps/antalmanac/src/stores/calendarizeHelpers.ts:
```diff
@@ -1,13 +1,18 @@
 
 const FINALS_WEEK_DAYS = ['Sat', 'Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri'];
 
+export function getBuilding(bldg: string): Location {
```
Same idea was one of my previous comments, can we keep the naming consistent so everything that previously referred to Building but now is actually a Location gets renamed properly?

EricPedley commented on apps/antalmanac/src/components/Map/Map.tsx:
```diff
@@ -249,9 +251,14 @@ export default function CourseMap() {
                 {/* Draw a marker for each class that occurs today. */}
                 {markersToDisplay.map((marker, index) => {
                     // Find all courses that occur in the same building prior to this one to stack them properly.
+                    // TODO Handle multiple buildings between class comparisons on markers.
```
Is this referring to the fact that we're only using `marker.bldg[0]` here and ignoring all others if they don't share the same `building` attribute?

ap0nia replied:
I don't remember making this comment, it might've been Alan. But yeah, we can only take one building because that's how the routing will calculate the route. When we show the routes, we show how to get from class 1 to class 2, class 2 to class 3, etc. and we don't have a way to represent class 1 to multiple locations for class 2.

EricPedley reviewed (changes requested):
Minor stuff, plus the big thing is renaming all stuff that used to just be a building string to now have "location" in the name.

EricPedley committed:
Replace `reduce` with flatmap

Co-authored-by: Aponia <brianmv@uci.edu>

alanchangxyz committed:
Syncing fork changes for multiple buildings (#696-a) (#697)

EricPedley reviewed (changes requested):
Ok basically looks fine, tested it out and everything works, just fix the failing tests and we're good here.

ap0nia committed:
fix: test cases

- the data received from PPA stays the same, i.e. bldg
- internally, we use use "locations" now

EricPedley reviewed (approved):
🚢🚢🚢🚢🚢
